cleverhans/picklable_model.py

```diff
@@ -0,0 +1,370 @@
+        self.kernels = PV(init)
+        if self.use_bias:
+            self.b = PV(np.zeros((self.output_channels,)).astype('float32'))
+        orig_input_batch_size = input_shape[0]
+        input_shape = list(input_shape)
+        input_shape[0] = 1
+        dummy_batch = tf.zeros(input_shape)
+        dummy_output = self.fprop(dummy_batch)
+        output_shape = [int(e) for e in dummy_output.get_shape()]
+        output_shape[0] = 1
+        self.output_shape = tuple(output_shape)
+
+    def fprop(self, x):
+        out = tf.nn.conv2d(x, self.kernels.var,
+                           (1,) + tuple(self.strides) + (1,), self.padding)
+        if self.use_bias:
+            out = out + self.b.var
+        return out
+
+    def get_params(self):
+        out = [self.kernels.var]
+        if self.use_bias:
+            out.append(self.b.var)
+        return out
+
+
+class ReLU(Layer):
+
+    def __init__(self, leak=0., **kwargs):
+        super(ReLU, self).__init__(**kwargs)
+        self.leak = leak
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        out = tf.nn.relu(x)
+        if self.leak != 0.0:
+            out = out - self.leak * tf.nn.relu(-x)
+        return out
+
+    def get_params(self):
+        return []
+
+
+class Sigmoid(Layer):
+
+    def __init__(self, **kwargs):
+        super(Sigmoid, self).__init__(**kwargs)
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        return tf.nn.sigmoid(x)
+
+    def get_params(self):
+        return []
+
+
+class Tanh(Layer):
+
+    def __init__(self, **kwargs):
+        super(Tanh, self).__init__(**kwargs)
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        return tf.nn.tanh(x)
+
+    def get_params(self):
+        return []
+
+
+class LeakyReLU(Layer):
+
+    def __init__(self, left_slope=.2, **kwargs):
+        super(LeakyReLU, self).__init__(**kwargs)
+        self.left_slope = left_slope
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        return tf.maximum(x, self.left_slope * x)
+
+    def get_params(self):
+        return []
+
+
+class ELU(Layer):
+
+    def set_input_shape(self, shape):
+        self.input_shape = shape
+        self.output_shape = shape
+
+    def get_output_shape(self):
+        return self.output_shape
+
+    def fprop(self, x):
+        return tf.nn.elu(x)
+
+    def get_params(self):
+        return []
+
+
+class SELU(Layer):
```

Comment:
I guess a few of these layers that just apply an activation function could inherit from the same class to save a few repeated lines but I'm not sure how much that would help/hurt readability.

Reply:
They're short enough that I mostly prefer to have duplicated code so that it's easy to change them later. For example, we might want to add a feature where you can learn the leak parameter of LeakyReLU, and then its get_params won't be an empty list anymore. We might want to add a feature where Softmax takes n - 1 inputs instead of n and then it's ouptut_shape isn't equal to its input_shape anymore, etc.